base/codevalidation.jl

```diff
@@ -0,0 +1,159 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+const VALID_EXPR_HEADS = Symbol[:call, :invoke, :static_parameter, :line, :gotoifnot, :(=),
+                                :method, :const, :null, :new, :return, :the_exception,
+                                :enter, :leave, :inbounds, :boundscheck, :copyast, :meta]
+
+const ASSIGNED_FLAG = 0x02
+
+struct InvalidCodeError <: Exception
+    errno::Int
+    msg::String
+    meta::Any
+end
+
+InvalidCodeError(errno, msg) = InvalidCodeError(errno, msg, nothing)
+
+"""
+    validate_code!(errors::Vector{>:InvalidCodeError}, c::CodeInfo)
+
+Validates `c`, logging any violations in `errors`.
+
+Any violation of the following properties will be logged by pushing a `InvalidCodeError`
+into `errors`:
+
+- `in(h, Base.Core.Inference.VALID_EXPR_HEADS)` for each subexpression head `h`
+- `Base.Core.Inference.is_valid_lhs(lhs)` for each encountered LHS value `lhs`
+- `Base.Core.Inference.is_valid_call_arg(arg)` for each call argument `arg`
+- `!isempty(c.slotnames)`
+- `length(c.slotnames) == length(c.slotflags)`
+- if `c.inferred`:
+    - `length(c.slottypes) == length(c.slotnames)`
+    - all SSAValues in AST have a type in `c.ssavaluetypes`
+- if `!c.inferred`:
+    - `c.slottypes == nothing`
+    - `c.ssavaluetypes` should be set to the number of SSAValues in `c.code`
+- all assigned-to slots have bit flag 2 set in their respective slotflags
+"""
+function validate_code!(errors::Vector{>:InvalidCodeError}, c::CodeInfo)
+    ssavals = SSAValue[]
+    walkast(c.code) do x
+        if isa(x, Expr)
+            if !in(x.head, VALID_EXPR_HEADS)
+                push!(errors, InvalidCodeError(1, "encountered invalid expression head", x.head))
+            elseif x.head == :(=) && !is_valid_lhs(x.args[1])
+                push!(errors, InvalidCodeError(2, "encountered invalid LHS value", x.args[1]))
```

Comment:
`is_valid_call_arg` can be renamed `is_valid_rvalue`, and also used to check the RHS of assignments here.